File: backend/src/models/order.ts
import { Schema, model } from "mongoose";

const orderSchema = new Schema({
  orderNumber: { type: String, unique: true }, // Human-readable order number
  buyerId: { type: Schema.Types.ObjectId, ref: "User", required: true },
  items: [
    {
      productId: {
        type: Schema.Types.ObjectId,
        ref: "Product",
        required: true,
      },
      sellerId: { type: Schema.Types.ObjectId, ref: "Seller" }, // Track which seller owns this item
      title: { type: String, required: true },
      image: { type: String },
      price: { type: Number, required: true },
      quantity: { type: Number, required: true },
      subtotal: { type: Number, required: true }, // price * quantity
    },
  ],
  totalPrice: { type: Number, required: true },
  platformFee: { type: Number, default: 0 }, // Platform commission
  sellerEarnings: { type: Number, default: 0 }, // Total earnings for sellers
  shippingAddress: {
    addressLine1: { type: String, required: true },
    addressLine2: { type: String },
    city: { type: String, required: true },
    postalCode: { type: String, required: true },
    country: { type: String, required: true },
    phoneNumber: { type: String },
    geolocation: {
      latitude: { type: Number },
      longitude: { type: Number },
    },
  },
  tracking: {
    currentLocation: {
      latitude: { type: Number },
      longitude: { type: Number },
      address: { type: String },
      updatedAt: { type: Date },
    },
    status: {
      type: String,
      enum: [
        "pending",
        "confirmed",
        "processing",
        "shipped",
        "out_for_delivery",
        "delivered",
        "canceled",
      ],
      default: "pending",
    },
    statusHistory: [
      {
        status: { type: String },
        timestamp: { type: Date, default: Date.now },
        location: { type: String },
        note: { type: String },
      },
    ],
    estimatedDelivery: { type: Date },
    trackingNumber: { type: String },
    carrier: { type: String }, // Shipping carrier name
  },
  paymentStatus: {
    type: String,
    enum: ["paid", "pending", "failed", "refunded"],
    default: "pending",
  },
  paymentMethod: {
    type: String,
    enum: ["paypal", "stripe", "cash"],
    default: "stripe",
  },
  paymentId: { type: String }, // Payment transaction ID
  refund: {
    status: {
      type: String,
      enum: ["none", "requested", "processing", "completed", "rejected"],
      default: "none",
    },
    amount: { type: Number, default: 0 },
    reason: { type: String },
    requestedAt: { type: Date },
    processedAt: { type: Date },
  },
  notes: { type: String }, // Order notes/instructions from buyer
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now },
});

// Pre-save hook to generate order number
orderSchema.pre("save", async function (next) {
  if (!this.orderNumber) {
    // Generate order number: ZMB-YYYYMMDD-XXXX
    const date = new Date();
    const dateStr = date.toISOString().slice(0, 10).replace(/-/g, "");
    const random = Math.floor(1000 + Math.random() * 9000);
    this.orderNumber = `ZMB-${dateStr}-${random}`;
  }
  next();
});

// Indexes for efficient querying
orderSchema.index({ buyerId: 1, createdAt: -1 });
orderSchema.index({ "tracking.status": 1 });
orderSchema.index({ "tracking.trackingNumber": 1 });
orderSchema.index({ orderNumber: 1 });
orderSchema.index({ "items.sellerId": 1 }); // For seller order queries

export const Order = model("Order", orderSchema);

File: frontend/zembil/lib/features/orders/domain/entity/order.dart
class OrderEntity {
  final String id;
  final List<OrderItem> items;
  final double totalPrice;
  final ShippingAddress shippingAddress;
  final OrderTracking tracking;
  final String paymentStatus;
  final String paymentMethod;
  final String? paymentId;
  final DateTime createdAt;
  final DateTime updatedAt;

  OrderEntity({
    required this.id,
    required this.items,
    required this.totalPrice,
    required this.shippingAddress,
    required this.tracking,
    required this.paymentStatus,
    required this.paymentMethod,
    this.paymentId,
    required this.createdAt,
    required this.updatedAt,
  });
}

class OrderItem {
  final String productId;
  final String title;
  final String? image;
  final double price;
  final int quantity;

  OrderItem({
    required this.productId,
    required this.title,
    this.image,
    required this.price,
    required this.quantity,
  });

  double get subtotal => price * quantity;
}

class ShippingAddress {
  final String addressLine1;
  final String? addressLine2;
  final String city;
  final String postalCode;
  final String country;
  final Geolocation? geolocation;

  ShippingAddress({
    required this.addressLine1,
    this.addressLine2,
    required this.city,
    required this.postalCode,
    required this.country,
    this.geolocation,
  });

  String get fullAddress {
    final parts = [
      addressLine1,
      if (addressLine2 != null && addressLine2!.isNotEmpty) addressLine2,
      city,
      postalCode,
      country,
    ];
    return parts.join(', ');
  }
}

class Geolocation {
  final double latitude;
  final double longitude;

  Geolocation({
    required this.latitude,
    required this.longitude,
  });
}

class OrderTracking {
  final TrackingLocation? currentLocation;
  final String status;
  final List<StatusHistory> statusHistory;
  final DateTime? estimatedDelivery;
  final String? trackingNumber;

  OrderTracking({
    this.currentLocation,
    required this.status,
    required this.statusHistory,
    this.estimatedDelivery,
    this.trackingNumber,
  });

  String get statusDisplay {
    switch (status) {
      case 'pending':
        return 'Order Pending';
      case 'confirmed':
        return 'Order Confirmed';
      case 'processing':
        return 'Processing';
      case 'shipped':
        return 'Shipped';
      case 'out_for_delivery':
        return 'Out for Delivery';
      case 'delivered':
        return 'Delivered';
      case 'canceled':
        return 'Canceled';
      default:
        return 'Unknown';
    }
  }

  bool get isActive => !['delivered', 'canceled'].contains(status);
  bool get canCancel => !['shipped', 'out_for_delivery', 'delivered', 'canceled'].contains(status);
}

class TrackingLocation {
  final double latitude;
  final double longitude;
  final String? address;
  final DateTime updatedAt;

  TrackingLocation({
    required this.latitude,
    required this.longitude,
    this.address,
    required this.updatedAt,
  });
}

class StatusHistory {
  final String status;
  final DateTime timestamp;
  final String? location;

  StatusHistory({
    required this.status,
    required this.timestamp,
    this.location,
  });
}

